approaches/abstract_approach.py

```diff
@@ -0,0 +1,47 @@
+import numpy as np
+import pandas as pd
+
+from text_models import AbstractModel
+from data_processing.util import get_corpus
+
+
+class AbstractApproach:
+    def __init__(self, train: pd.DataFrame, test: pd.DataFrame):
+        self.train = train
+        self.test = test
+        self.train_corpus = get_corpus(train)
+        self.test_corpus = get_corpus(test)
+
+        self.embeddings = None
+        self.test_embs = None
+
+    def evaluate(self, model: AbstractModel, topn=5):
+        self.embeddings = model.get_embeddings(self.train_corpus)
+        self.test_embs = model.get_embeddings(self.test_corpus)
+
+        self.additional_preparation()
+
+        test_size = 0
+        TP = 0
+        for ind in range(len(self.test_corpus)):
+            if self.test.iloc[ind]["id"] != self.test.iloc[ind]["disc_id"]:  # not in master_ids
```

Comment:
This kind of iteration over a large DataFrame can be incredibly slow as you basically ask DataFrame for random access while in fact iterating over it. DataFrames have a lot of built-in functions  that you can use instead. Here you can likely vectorize almost everything and make the code way faster and cleaner. Look at functions like `DF.map`, `DF.apply`, etc.

A [post](https://towardsdatascience.com/how-to-make-your-pandas-loop-71-803-times-faster-805030df4f06) explaining what's going on.

Reply:
Now instead of the loop I use `DF.apply`